Applications/util/mkfs.8
MKFS(8)
## NAME
*mkfs* \[-f\] device isize fsize

## DESCRIPTION
The *mkfs* tool initializes a new file system on a device.

The isize field specifies the number of inode blocks on the file system.
Each inode block holds 8 inodes. This determines the number of individual
files and directories that may be present on the disk.

The fsize field gives the size of the media in 512 byte blocks - up to 65535
blocks for 32MB media.

The default behaviour is to zero the entire file system and then write the
new metadata to the device. If the -f flag is passed then the data areas are
not cleared, bad data blocks may not be found and old data may persist in
blocks not allocated to files.

## BUGS
None know

## STANDARDS
Whilst most Fuzix-like systems have a mkfs command the syntax, functionality
and file system format varies extensively.


RELEASE-NOTES

			Fuzix 0.2 Release Notes

Fuzix 0.2 is the second major release of the Fuzix OS. The primary focus has
been completing the core C library functionality. Applications have also been
updated and the kernel has a considerable number of bugs fixed and new
platforms.

Supported Hardware

Amstrad NC100 with 1MB SRAM card
Amstrad NC200 with 1MB SRAM card and floppy disk drive
Dragon 32/64 with Spinx or MOOH cartridge
DX Designs P112 with 1MB RAM and G-IDE
EACA Video Genie (with suitable banked memory)
LNW Research LNW80 with suitable banked memory (TRS80 Model 1 compatible bits)
Memotech MTX512 with SDX disk (no Rememorizer or Rememotech support)
Microbee 256TC or Premium with 128K+ RAM and ideally hard disk
MSX2 + MegaFlashROM with SD
Multicomp09 with at least 128K RAM
RBC(*) Mark IV with optional PropIO V2
RBC(*) SBC v2 with optional PPIDE and/or PropIO V2
RC2014 SBC with 512K ROM/RAM, CF, SIO and RTC
RC2014 SBC with banked ROM, 64K RAM, CF, SIO and RTC
SOCZ80 (128MHz FPGA Z80 platform)
Tandy COCO2 with Cloud 9 IDE, or COCOSDC card and Fuzix partly in cartridge
Tandy COCO3 with suitable disk interface
TRS80 Model I/III with a hard disk and a supported banked RAM expansion
TRS80 Model 4/4D/4P 128K RAM
Zeta
Zeta V2

(*) Formerly N8VEM now RBC (RetroBrewComputers)

Virtual Platforms (Emulation Platforms For Fun or Development)

V65C816		-	65C816 emulation work
V68		-	68000 development work
Z80Pack		-	Very flexible Z80 banked system emulator

Obsolete/Dropped Platforms

MSP430FR5969 - bitrotted
MSX1 with MegaMem - bitrotted
TGL6502 - project died
ZX128 (Spectrum 128K etc) - too difficult to make work well

Incomplete Ports

Amstrad PCW8256/8512/9256/9512/10
Apple IIc
Cromemco
Dragon 32/64 MOOH (full MOOH support, use the Spinx/Mooh port for now)
Gemini
Lucas Nascom
Sam Coupe
SC108 (port complete for proposed hardware)
V65 (Virtual 6502 development platform)
Video Genie with EG64
Z280RC


The Library/Application support has been built for
6502			no FP
65C816			no FP, using 6502 modes
6809			FP needs debug
68000			FP needs debug
68HC11			test only
8086			test only
MSP430X
PDP11			test only
Rabbit2000		test only
Z80
Z180



Changes From 0.2 to 0.2.1

o	Fixed a bug where zombies got swapped in error. Fixes the TRS80
	hang at the date prompt
o	Fixed a warning about banking fixups on the Model1
o	Better debugging in swap.c
o	Better debugging in binmunge
o	Remembered to add the release notes!
o	Spell MOOH correctly


Applications/util/mkfs.c
/**************************************************
UZI (Unix Z80 Implementation) Utilities:  mkfs.c
***************************************************/

/* This utility creates an UZI file system for a defined block device.
 * the format is:
 *                  mkfs device isize fsize
 * where: device is the logical block on which to install a filesystem
 *        isize is the number of 512-byte blocks (less two reserved for
 *            system use) to use for storing 64-byte i-nodes
 *        fsize is the total number of 512-byte blocks to assign to the
 *            filesystem.  Space available for storage of data is therefore
 *            fsize-isize blocks.
 *
 * Revisions:
 *   Based on UZI280 version (minor changes from original UZI.  HFB
 *   Modified for use under UZI.                                HP
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/super.h>


struct dinode {
    uint16_t i_mode;
    uint16_t i_nlink;
    uint16_t i_uid;
    uint16_t i_gid;
    off_t    i_size;
    uint32_t   i_atime;		/* Breaks in 2038 */
    uint32_t   i_mtime;		/* Need to hide some extra bits ? */
    uint32_t   i_ctime;		/* 24 bytes */
    blkno_t  i_addr[20];
};               /* Exactly 64 bytes long! */

#define FILENAME_LEN	30
#define DIR_LEN		32
typedef struct direct {
    uint16_t   d_ino;
    char     d_name[FILENAME_LEN];
} direct;


uint8_t fast=0;     /* flag for fast formatting option */

int dev;

#define ROOTINODE 1       /* Inode # of / for all mounted filesystems. */

direct dirbuf[64] = { {ROOTINODE, "."}, {ROOTINODE, ".."} };
struct dinode inode[8];
struct fuzix_filesys_kernel fs_tab;

void dwrite(uint16_t blk, char *addr)
{
    if (lseek(dev, blk * 512L, 0) == -1) {
        perror("lseek");
        exit(1);
    }
    if (write(dev, addr, 512) != 512) {
        perror("write");
        exit(1);
    }
}

char *zerobuf(void)
{
    static char buf[512];
    
    memset(buf, 0, 512);
    return buf;
}


int yes(void)
{
    char line[20];

    fflush(stdout);

    if (!fgets(line, sizeof(line), stdin) || (*line != 'y' && *line != 'Y'))
	return (0);

    return (1);
}

void mkfs(uint16_t fsize, uint16_t isize)
{
    uint16_t j;
    char *zeros;

    /* Zero out the blocks */
    printf("Zeroizing i-blocks...\n");
    zeros = zerobuf();		/* Get a zero filled buffer */

    if( !fast ){
	    for (j = 0; j < fsize; ++j)
		    dwrite(j, zeros);
    }
    else{
	    for (j = 0; j < isize; ++j)
		    dwrite(j, zeros);
    }

    /* Initialize the super-block */
    fs_tab.s_mounted = SMOUNTED;	/* Magic number */
    fs_tab.s_isize = isize;
    fs_tab.s_fsize = fsize;
    fs_tab.s_nfree = 1;
    fs_tab.s_free[0] = 0;
    fs_tab.s_tfree = 0;
    fs_tab.s_ninode = 0;
    fs_tab.s_tinode = (8 * (isize - 2)) - 2;

    /* Free each block, building the free list */

    printf("Building free list...\n");
    for (j = fsize - 1; j > isize; --j) {
	if (fs_tab.s_nfree == 50) {
	    dwrite(j, (char *) &fs_tab.s_nfree);
	    fs_tab.s_nfree = 0;
	}
	++fs_tab.s_tfree;
	fs_tab.s_free[(fs_tab.s_nfree)++] = j;
    }

    /* The inodes are already zeroed out */
    /* create the root dir */

    inode[ROOTINODE].i_mode = S_IFDIR | 0777;
    inode[ROOTINODE].i_nlink = 3;
    inode[ROOTINODE].i_size = 64;
    inode[ROOTINODE].i_addr[0] = isize;

    /* Reserve reserved inode */
    inode[0].i_nlink = 1;
    inode[0].i_mode = ~0;

    printf("Writing initial inode and superblock...\n");

    sync();
    dwrite(2, (char *) inode);
    dwrite(isize, (char *) dirbuf);

    sync();
    /* Write out super block */
    dwrite(1, (char *) &fs_tab);

    sync();
    printf("Done.\n");
}

void printopts(void)
{
	fprintf( stderr, "usage: mkfs [options] device isize fsize\n");
	exit(-1);
}


int main(int argc, char *argv[])
{
    uint16_t fsize, isize;
    struct stat statbuf;
    int option;

    while( (option=getopt( argc, argv, "f" ))>0 ){
	    switch( option ){
	    case 'f':
		    fast=1;
		    break;
	    case '?':
	    default:
		    printopts();
	    }
    }

    if (argc-optind < 3) printopts();
    
    
    if (stat(argv[optind], &statbuf) != 0) {
        fprintf(stderr, "mkfs: can't stat %s\n", argv[optind]);
        exit(-1);
    }

    if (!S_ISBLK(statbuf.st_mode)) {
        fprintf(stderr, "mkfs: %s is not a block device\n", argv[optind]);
        exit(-1);
    }

    isize = (uint16_t) atoi(argv[optind+1]);
    fsize = (uint16_t) atoi(argv[optind+2]);

    if (fsize < 3 || isize < 2 || isize >= fsize) {
	fprintf(stderr, "mkfs: bad parameter values\n");
	exit(-1);
    }

    printf("Making filesystem on device %s with isize %u fsize %u. Confirm? ",
	   argv[optind], isize, fsize);
    if (!yes())
	exit(-1);

    dev = open(argv[optind], O_RDWR|O_SYNC);
    if (dev < 0) {
        fprintf(stderr, "mkfs: can't open device %s\n", argv[optind]);
        exit(-1);
    }

    mkfs(fsize, isize);

    exit(0);
}



